File: src/client/java/fr/hardel/asset_editor/client/highlight/HighlightPalette.java
package fr.hardel.asset_editor.client.highlight;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public final class HighlightPalette {

    private final LinkedHashMap<String, HighlightStyle> styles = new LinkedHashMap<>();
    private final List<Runnable> listeners = new ArrayList<>();

    public HighlightPalette set(String name, HighlightStyle style) {
        Objects.requireNonNull(name, "name");
        Objects.requireNonNull(style, "style");
        styles.put(name, style);
        notifyListeners();
        return this;
    }

    public HighlightStyle get(String name) {
        return styles.get(name);
    }

    public boolean contains(String name) {
        return styles.containsKey(name);
    }

    public boolean delete(String name) {
        boolean removed = styles.remove(name) != null;
        if (removed)
            notifyListeners();

        return removed;
    }

    public void clear() {
        if (styles.isEmpty())
            return;

        styles.clear();
        notifyListeners();
    }

    public Map<String, HighlightStyle> snapshot() {
        return Map.copyOf(styles);
    }

    public void addListener(Runnable listener) {
        listeners.add(listener);
    }

    public void removeListener(Runnable listener) {
        listeners.remove(listener);
    }

    private void notifyListeners() {
        for (Runnable listener : List.copyOf(listeners))
            listener.run();
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/lib/highlight/HighlightPalette.kt
package fr.hardel.asset_editor.client.compose.lib.highlight

import java.util.LinkedHashMap
import java.util.Objects
import java.util.concurrent.CopyOnWriteArrayList

class HighlightPalette {

    private val styles = LinkedHashMap<String, HighlightStyle>()
    private val listeners = CopyOnWriteArrayList<Runnable>()

    fun set(name: String, style: HighlightStyle): HighlightPalette {
        Objects.requireNonNull(name, "name")
        Objects.requireNonNull(style, "style")
        styles[name] = style
        notifyListeners()
        return this
    }

    fun get(name: String): HighlightStyle? = styles[name]

    fun contains(name: String): Boolean = styles.containsKey(name)

    fun clear() {
        if (styles.isEmpty()) {
            return
        }
        styles.clear()
        notifyListeners()
    }

    fun snapshot(): Map<String, HighlightStyle> = styles.toMap()

    fun addListener(listener: Runnable) {
        listeners.add(listener)
    }

    fun removeListener(listener: Runnable) {
        listeners.remove(listener)
    }

    private fun notifyListeners() {
        listeners.forEach(Runnable::run)
    }
}
